src/OpenTelemetry/Metrics/AggregatorStore.cs

```diff
@@ -217,18 +231,56 @@ private int LookupAggregatorStore(string[] tagKeys, object[] tagValues, int leng
             return aggregatorIndex;
         }
 
+        private bool TryGetUnusedMetricPoint(out int index)
+        {
+            if (this.metricPointIndex < MaxMetricPoints)
+            {
+                index = Interlocked.Increment(ref this.metricPointIndex);
+                if (index < MaxMetricPoints)
+                {
+                    return true;
+                }
+            }
+
+            if (this.temporality == AggregationTemporality.Delta)
+            {
+                for (int i = 1; i < MaxMetricPoints; i++)
+                {
+                    ref var metricPoint = ref this.metricPoints[i];
+                    if (metricPoint.MetricPointStatus == MetricPointStatus.CandidateForRemoval)
+                    {
+                        if (metricPoint.TryFree())
+                        {
+                            // Concurrency issue here.
+                            // LookupAggregatorStore may have retrieved the MP we just freed.
```

Comment:
yes. the lookup + update is not atomic is the core issue that needs addressing.
Some attempts to take RWLslim for dictionary look up: 
https://github.com/open-telemetry/opentelemetry-dotnet/compare/main...cijothomas/readwritelockdictionary